banyan/src/forest/read.rs

```diff
@@ -19,7 +13,260 @@ use anyhow::{anyhow, Result};
+        Self {
+            index,
+            position: 0,
+            filter: smallvec![],
+        }
+    }
+    fn is_exhausted(&self, mode: &Mode) -> bool {
+        match mode {
+            Mode::Forward => !self.filter.is_empty() && self.position == self.filter.len() as isize,
+            Mode::Backward => self.position < 0,
+        }
+    }
+    fn next_pos(&mut self, mode: &Mode) {
+        match mode {
+            Mode::Forward => self.position += 1,
+            Mode::Backward => self.position -= 1,
+        }
+    }
+}
+
+impl<T: TreeTypes, V, R, Q, E, F> ForestIter<T, V, R, Q, F>
+where
+    T: TreeTypes + 'static,
+    V: DagCbor + Clone + Send + Sync + Debug + 'static,
+    R: ReadOnlyStore<T::Link> + Clone + Send + Sync + 'static,
+    Q: Query<T> + Clone + Send + 'static,
+    E: Send + 'static,
+    F: Fn(IndexRef<T>) -> E + Send + Sync + 'static,
+{
+    pub(crate) fn new(
+        forest: Forest<T, V, R>,
+        query: Q,
+        index: Arc<Index<T>>,
+        mk_extra: F,
+    ) -> Self {
+        let mode = Mode::Forward;
+        let stack = smallvec![TraverseState::new(index)];
+
+        Self {
+            forest,
+            offset: 0,
+            query,
+            mk_extra,
+            stack,
+            mode,
+        }
+    }
+    pub(crate) fn new_rev(
+        forest: Forest<T, V, R>,
+        query: Q,
+        index: Arc<Index<T>>,
+        mk_extra: F,
+    ) -> Self {
+        let offset = index.count();
+        let mode = Mode::Backward;
+        let stack = smallvec![TraverseState::new(index)];
+
+        Self {
+            forest,
+            offset,
+            query,
+            mk_extra,
+            stack,
+            mode,
+        }
+    }
+}
+
+impl<T: TreeTypes, V, R, Q, E, F> Iterator for ForestIter<T, V, R, Q, F>
+where
+    T: TreeTypes + 'static,
+    V: DagCbor + Clone + Send + Sync + Debug + 'static,
+    R: ReadOnlyStore<T::Link> + Clone + Send + Sync + 'static,
+    Q: Query<T> + Clone + Send + 'static,
+    E: Send + 'static,
+    F: Fn(IndexRef<T>) -> E + Send + Sync + 'static,
+{
+    type Item = Result<FilteredChunk<T, V, E>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let res: FilteredChunk<T, V, E> = loop {
+            let head = match self.stack.last_mut() {
+                Some(i) => i,
+                // Nothing to do ..
+                _ => return None,
+            };
+
+            //  Branch is exhausted: Ascend.
+            if head.is_exhausted(&self.mode) {
+                // Ascend to parent's node
+                self.stack.pop();
+
+                // increase last stack ptr, if there is still something left to
+                // traverse
+                if !self.stack.is_empty() {
+                    let last = self.stack.last_mut().expect("not empty");
+                    last.next_pos(&self.mode);
+                }
+                continue;
+            }
+
+            match self.forest.load_node(&head.index) {
```

Comment:
no, it is indeed a bit weird. All nodes in banyan are "split in half". It is not very object oriented, since data that you would logically group together is separate so it is ordered by probability / frequency of access.